Update for Swift 6.1 hash algorithm change

luispadron opened the pull request:
Fixes #120. Swift 6.1 changes the hash algorithm to BLAKE3, this change is not backwards compatible as we need to build with swiftlang/llvm-project (Swift 6.1 release). Projects that need to support all Swift 6.x (Xcode 16.0-16.2) can use both binaries (before and after this PR).

keith commented:
lmk when i should review, lgtm so far once you remove the debugging stuff

luispadron commented:
@keith tested end-to-end works with Xcode 16.3 and double checked the flag value can be toggled on/off

keith commented:
can you pull this in https://github.com/MobileNativeFoundation/index-import/pull/132, i think we'll need to do them in lockstep

luispadron commented:
> can you pull this in #132, i think we'll need to do them in lockstep

I updated the code (was doing the flag stuff weirdly, it now matches the other flags). I also added a commit to update CI to use Swift 6.1

keith commented:
i think you'll have to update the tests too https://github.com/MobileNativeFoundation/index-import/blob/919db68ede9c253a68bd5c496320b219cb76f426/tests/run.sh#L34

actually a good chance to add a new condition with the old hash names + your flag

luispadron commented:
> i think you'll have to update the tests too
> 
> https://github.com/MobileNativeFoundation/index-import/blob/919db68ede9c253a68bd5c496320b219cb76f426/tests/run.sh#L34
> 
> actually a good chance to add a new condition with the old hash names + your flag

Can look at updating tests tomorrow! Wouldnt we need to have another job thats not using Swift 6.1 to test the legacy hashing flag?

luispadron commented:
@keith Updated CI & test for 6.1, also updated the test script so in the future we can test with the old hash algo using `--use-legacy-hashing`. Can do that in another PR maybe?

keith commented:
> Can look at updating tests tomorrow! Wouldnt we need to have another job thats not using Swift 6.1 to test the legacy hashing flag?

since the same binary can produce both, and the tests themselves just test the production, not the consumption, of the files, i think we should be able to make the same run script run it twice, once in both modes, to test them

keith commented:
a diff change is fine with me, looks like there's still a failure here

luispadron commented:
It's expecting the old hash but I thought this should be the new hash since its running with Swift 6.1? Is the setup there not right, maybe its Xcode 15.4 from the macos-14 machines using an older Swift version?

keith commented:
this is coming from test fixtures https://github.com/MobileNativeFoundation/index-import/blob/919db68ede9c253a68bd5c496320b219cb76f426/tests/clang/expected.txt#L17, there are a few

luispadron commented:
> this is coming from test fixtures
> 
> https://github.com/MobileNativeFoundation/index-import/blob/919db68ede9c253a68bd5c496320b219cb76f426/tests/clang/expected.txt#L17
> 
> , there are a few

hmm yah but i updated those expected files to use the new hash and CI should be using Swift/llvm 6.1 which should be producing the new hash. This works locally for me, just failing in github ci.

keith commented:
i bet it's using the wrong version of xcode? https://github.com/MobileNativeFoundation/index-import/blob/e0a3cbdc52423707a81a806f13711482c631383b/tests/run.sh#L10-L12

keith commented on tests/run.sh:
```diff
@@ -122,7 +128,7 @@ clang -fsyntax-only -index-store-path input input.c
 
 # Check that the expected index files exist.
-ls output/v5/units/output.c.o-1I92L511L7IRP >/dev/null
+ls output/v5/units/output.c.o-2LQD3ZSM9CGHD >/dev/null
```
why did these change?

luispadron commented:
Had to disable the import only tests because index-import now expects the new hashes and we cant use Xcode 16.3 in GitHub CI

keith commented on tests/run.sh:
```diff
@@ -41,64 +41,68 @@ popd >/dev/null
 
-echo "Testing import-output-file"
-pushd "$base_dir"/import_only >/dev/null
+# TODO: enable this test again when using CI with Xcode 16.3+
```
can you fix the tests now to verify they work locally as we'd expect

luispadron replied:
hm from what i understood these tests are failing _in_ index-import with:

```
Testing import-output-file
error: failed to read unit file input/v5/units/input.c.o-3TYUTGSZ8FF3H -- Failed opening 'input/v5/units/input.c.o-3TYUTGSZ8FF3H': No such file or directory
```

I think this is because the has expects the Swift 6.1 one but CI is using Xcode 15.x. Is it possible to fix these, GitHub doesnt have Xcode 16.3 available.

See: https://github.com/MobileNativeFoundation/index-import/actions/runs/14232916900/job/39886942224

keith replied:
yea i mean fix the script and disable the github action, so we can at least test locally

luispadron replied:
ahhh yeah good idea, let me do that

luispadron replied:
Tests pass locally now when using Xcode 16.3, disabled it in CI until github supports it

Wilfred commented:
Thanks for fixing this!

To confirm my understanding: you decided to support a single version of the hash in the final version of the PR, there's no CLI flag to toggle the hash used. Seems reasonable :)

keith commented:
yea because the APIs in clang we're calling use the same hash logic, so we couldn't put that behind a flag (without trying to dynamically link the lib from the current xcode or something)
